src/sage/algebras/quatalg/quaternion_algebra.py

```diff
@@ -2870,30 +2938,142 @@ def is_equivalent(self, J, B=10) -> bool:
+                                    'for indefinite quaternion algebras')
 
-        # Just test theta series first.  If the theta series are
-        # different, the ideals are definitely not equivalent.
+        # Just test theta series first; if the theta series are
+        # different, the ideals are definitely not equivalent
         if B > 0 and self.theta_series_vector(B) != J.theta_series_vector(B):
+            if certificate:
+                return False, None
             return False
 
         # The theta series are the same, so perhaps the ideals are
-        # equivalent.  We use Prop 1.18 of [Pizer, 1980] to decide.
+        # equivalent.  We use Prop 1.18 of [Piz1980]_ to decide.
         # 1. Compute I * Jbar
-        # see Prop. 1.17 in Pizer.  Note that we use IJbar instead of
-        # JbarI since we work with right ideals
+        # Note that we use I*Jbar instead of Jbar*I since we work with right ideals
         IJbar = self.multiply_by_conjugate(J)
 
-        # 2. Determine if there is alpha in K such
-        #    that N(alpha) = N(I)*N(J) as explained by Pizer.
+        # 2. Determine if there is alpha in K such that
+        # N(alpha) = N(I)*N(J) as explained by Pizer
         c = IJbar.theta_series_vector(2)[1]
-        return c != 0
+        if c == 0:
+            if certificate:
+                return False, None
+            return False
+
+        if not certificate:
+            return True
+
+        # Use Pari's qfminim with flag = 1 to find alpha as above
+        _,v = IJbar.quadratic_form().__pari__().qfminim(None, None, 1)
+        alpha = sum(ZZ(c)*g for c,g in zip(v, IJbar.basis()))
+
+        # As explained by Pizer, we now only need to rescale alpha by 1/N(J)
+        return True, alpha * (1/J.norm())
+
+    def is_principal(self, certificate=False):
+        r"""
+        Checks whether ``self`` is principal as a full rank quaternion ideal.
+        Requires the underlying quaternion algebra to be definite.
+        Independent of whether ``self`` is a left or a right ideal.
+
+        INPUT:
+
+        - ``certificate`` -- if ``True`` returns a generator alpha s.t. I = alpha*O
+          where O is the right order of I.
+
+        OUTPUT: bool, or (bool, alpha) if ``certificate`` is ``True``
+
+        EXAMPLES::
+
+            sage: B.<i,j,k> = QuaternionAlgebra(419)
+            sage: O = B.quaternion_order([1/2 + 3/2*j, 1/6*i + 2/3*j + 1/2*k, 3*j, k])
+            sage: alpha = -1 - 59*i + 1/2*j + 154/9*k
+            sage: I = alpha * O
+            sage: I.is_principal(True)
+            (True, 1 + 59*i - 1/2*j - 154/9*k)
+        """
+        if not self.quaternion_algebra().is_definite():
+            raise NotImplementedError('principality test not'
+                                    'implemented in indefinite'
+                                    'quaternion algebras')
+
+        if len(self.basis()) < 4:
```

Comment:
Can this fail?

Reply:
Interesting, I was expecting the constructor for fractional ideals to check the rank. That should probably be fixed eventually, but it doesn't necessarily have to be done here and now.